cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,139 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
                      programTerminator, ownerAdmin.getOwner(appId), appRequest.canUpdateSchedules());
   }
 
+  /**
+   * Upgrades an existing application by upgrading appliation artifact versions and plugin artifact versions.
+   *
+   * @param appId the id of the application to upgrade.
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @return Details of upgrade for the application like failure details with reason.
+   * @throws IllegalStateException if something unexpected happened during upgrade.
+   * @throws IOException if there was an IO error during initializing application class from artifact.
+   * @throws JsonIOException if there was an error in serializing or deserializing app config.
```

Comment:
is this the right exception?

Reply:
Comes from toJSON method used during json serialization in updateInternal method.